src/kernel/boot.c

```diff
@@ -623,12 +597,49 @@ BOOT_CODE static bool_t create_untypeds_for_region(
          * be used anyway.
          */
         if (size_bits >= seL4_MinUntypedBits) {
-            if (!provide_untyped_cap(root_cnode_cap, device_memory, reg.start, size_bits, first_untyped_slot)) {
+
+            /* The number of caps that can be stored in the boot info is fixed,
+             * the system should be set up that all caps can be stored there. We
+             * do not consider it a fatal error if there is no enough space,
+             * because it just means we can't provide untyped capabilities for
+             * some memory areas. The system might still be able to run if it
+             * does not use all system resources.
+             */
+            word_t idx = ndks_boot.slot_pos_cur - first_untyped_slot;
+            if (idx >= CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS) {
+                printf("WARNING: can't provide cap for %"SEL4_PRIx_word"/2^%u"
+                       " and beyond, CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS (%d)"
+                       " exceeded\n",
+                       reg.start, size_bits,
+                       (int)CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS);
+                assert(0); /* Make this fatal for debug builds only. */
```

Comment:
Not sure it should be fatal for debug either. In fact, it might be especially in debug mode where you want the warning but otherwise want things to proceed while you are bringing up something new.

Reply:
It might make sense to add to the warning that we're proceeding anyway and that this means the memory will not be available to user space.

Reply:
I'm ok with removing this assert if nobody thing that it should be kept. It's just there from my local patch set where this was useful to catch some issues early. If we remove this assert, the warning can show up multiple times when the configuration is tool small - which can get a bit annoying if this configuration is intended.

Reply:
remove the assert